### PR TITLE
oop-practice. Repaired output. Using array of objects and forEach()

### PR DESCRIPTION
**Pull request by LGeorge007**
### OOP-practice

[Demo](https://lgeorge007.github.io/oop-practice)
[Code base](https://github.com/LGeorge007/oop-practice)

The code is submitted in a dedicated feature branch.
Please, review.

### CODE REVIEW

**Review by OleksiyRudenko — CHANGES_REQUESTED** [review]
@LGeorge007 good work.
However some of the inconsistencies make me think you were in hurry.
Take your time and re-check your code for consitency.

Please also make your IDE to add empty lines at a file end for you.

image

File: submissions/LGeorge007/oop-practice/index.js [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+   constructor(name, species, gender, greeting, friends) {
+       this.species = species;
+       this.name = name;
+       this.gender = gender;
+       this.greeting = greeting;
+       this.friends = friends;
+   }
+  
+   say() {
+       let haveFriend = (this.friends !== undefined) ? (this.friends.length === 1) ? "I have a friend. This is " + this.friends[0] : "I have friends. These are "+ this.friends.join(", ")   : "I'm alone. I have no friends!";
```
[the remark this answers is not in the document]
Comment (LGeorge007) [reply submissions/LGeorge007/oop-practice/index.js]: I've fixed my VS Code. And now it added new empty line in the end of file.
“files.insertFinalNewline”: true,
“files.trimFinalNewlines”: true,
“files.trimTrailingWhitespace”: true

I don't hear about it. But It was found of more convinient.
 let haveFriend = (this.friends !== undefined)
         ? (this.friends.length === 1)
            ? "I have a friend. This is " + this.friends[0]
            : "I have friends. These are "+ this.friends.join(", ")
         : "I'm alone. I have no friends";

File: submissions/LGeorge007/oop-practice/index.js [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+}
+
+class Animal extends Inhabitant {
+   constructor( name, species, gender, greeting, friends) {
+      super( name, species, gender, greeting, friends)
+      this.paws = 4;
+   }
+
+   say() {
+       return `${this.greeting} ` + super.say() + ` I have ${this.paws} paws.`
```
[the remark this answers is not in the document]
Comment (LGeorge007) [reply submissions/LGeorge007/oop-practice/index.js]: I defined greeting in the Super class.
And use it in  say() function.
return `${this.greeting} I'm ${this.name}. I'm a ${this.species} and I have a ${this.gender} gender. ${haveFriend}.`;

File: submissions/LGeorge007/oop-practice/index.js [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+   }
+}
+
+class Human extends Inhabitant {
+   constructor(name, species, gender,  greeting, friends) {
+      super(name, species, gender, greeting, friends);
+      this.legs = 2;
+      this.hands = 2;
+      this.greeting = greeting;
```
[the remark this answers is not in the document]
Comment (LGeorge007) [reply submissions/LGeorge007/oop-practice/index.js]: Yes. I agreed with You. 
It does not make sense.
I defined greeting in the Inhabitant class.
And use it in say() function() there.

File: submissions/LGeorge007/oop-practice/index.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (OleksiyRudenko) [reply submissions/LGeorge007/oop-practice/index.js]: > **I don't hear about it.** But It was found of more convinient.

What do you mean?

[the remark this answers is not in the document]
Comment (LGeorge007) [reply submissions/LGeorge007/oop-practice/index.js]: This spelling is more convenient for reading. 
condition
  ? a
    ? x
    : v
  : t
  I haven't used this spelling before.

File: submissions/LGeorge007/oop-practice/index.js [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+}
+
+let catVoice = "Meow-meow!";
+
+const inhabitantsList = [ new Human("Boris", "human", "male", "HI!", ["Vladimir", "Innokentiy", "Fedor"]),
+                       new Human("Larisa", "human", "female", "Hello!", ["Diana","Laura","Svetlana"]),
+                       new Dog("Sharikk", "dog", "male", "Woof-woof!", ["Bublik"]),
+                       new Cat("Murzik", "cat", "male",  catVoice, ["Sonya", "Garfild", "Carapka"]),
+                       new CatWoman ("Selina Kyle", "humicat", "female", catVoice, )
+                     ];
```
Comment (OleksiyRudenko) [inline submissions/LGeorge007/oop-practice/index.js:68]: Every class of `Human`, `Dog`, `Cat`, `CatWoman` expect species explicitly passed to them.
It's unlikely we want a user be able to create a human with a different species. Same is tru fro dogs, cats and catwomen.
Why not letting `Human`, `Dog`, `Cat`, `CatWoman` classes to handle species by themselves?
They can pass respective species literals to their supers, expecting no species in their constructors.

**Review by OleksiyRudenko — CHANGES_REQUESTED** [review]
@LGeorge007 nearly done.
One thing yet to fix.

**Review by OleksiyRudenko — APPROVED** [review]
@LGeorge007 well done!

**Pull request merged**